src/components/FlyoutMenu/FlyoutMenu.js

```diff
@@ -0,0 +1,154 @@
+  /** The button that triggered the menu. */
+  trigger: PropTypes.oneOf(TRIGGERS).isRequired,
+  /** The menu content. */
+  children: PropTypes.node.isRequired,
+  /** Fired when the menu is closed. */
+  onClose: PropTypes.func,
+};
+
+function FlyoutMenu({ menu, trigger, onClose, children }) {
+  const dispatch = useDispatch();
+
+  const allOtherMenus = useMemo(() => MENUS.filter(m => m !== menu), [menu]);
+
+  const overlayRef = useRef();
+  const getFocusableElements = useCallback(() => {
+    return overlayRef.current.querySelectorAll(focusableElementDomString);
+  }, []);
+
+  const isDisabled = useSelector(state => selectors.isElementDisabled(state, menu));
+  const isOpen = useSelector(state => selectors.isElementOpen(state, menu));
+
+  const closeMenu = useCallback(() => {
+    dispatch(actions.closeElements([menu]));
+    onClose && onClose();
+  }, [dispatch, menu, onClose]);
+
+  const onClickOutside = useCallback(
+    e => {
+      const menuButton = document.querySelector(`[data-element="${trigger}"]`);
+      const clickedMenuButton = menuButton?.contains(e.target);
+      if (!clickedMenuButton) {
+        closeMenu();
+      }
+    },
+    [closeMenu, trigger],
+  );
+  useOnClickOutside(overlayRef, onClickOutside);
+
+  const [position, setPosition] = useState(() => ({ left: 0, right: 'auto', top: 'auto' }));
+  const isMobile = useMedia(['(max-width: 640px)'], [true], false);
+  const isTabletOrMobile = useMedia(['(max-width: 900px)'], [true], false);
+
+  // When open: close others, position, and listen for resizes to position.
+  useEffect(() => {
+    if (isOpen) {
+      dispatch(actions.closeElements(allOtherMenus));
+
+      const onResize = () => setPosition(getOverlayPositionBasedOn(trigger, overlayRef, isTabletOrMobile));
+      onResize();
+
+      window.addEventListener('resize', onResize);
+      return () => window.removeEventListener('resize', onResize);
+    }
+  }, [allOtherMenus, dispatch, isOpen, isTabletOrMobile, trigger]);
+
+  // When menu opens, focus will be controlled internally with Up/Down arrows.
+  // Will return focus to button that opened it once it closes.
+  useEffect(() => {
+    if (!isDisabled && isOpen) {
+      const lastFocusedElement = document.activeElement;
+
+      const keydownListener = e => {
+        // Tab closes the menu.
+        if (e.key === 'Tab' || e.key === 'Escape') {
+          e.preventDefault();
+          closeMenu();
+          if (lastFocusedElement) {
+            lastFocusedElement.focus();
+          }
+          return;
+        }
+
+        // Up and Down keys used to navigate overlay menu.
+        if (e.key === 'ArrowUp' || e.key === 'ArrowDown') {
+          const focusable = getFocusableElements();
+
+          if (focusable.length === 0) {
+            return;
+          }
+
+          const focusIndex = findFocusableIndex(focusable, document.activeElement);
+
+          if (e.key === 'ArrowUp') {
+            if (focusIndex <= 0) {
```

Comment:
I think the logic here can be simplified to
```js
let nextIndex;
if (e.key === 'ArrowUp') {
  nextIndex = (focusIndex - 1 + length) % length;
} else {
  nextIndex = (focusIndex + 1 + length) % length;
}

focusable[nextIndex].focus();
```
(if we want to be fancy （￣︶￣）↗　 )

Reply:
Good call, gonna do that